### PR TITLE
SFB-81: Implement feature flags

### PR DESCRIPTION
**Pull request by JonasVanHoof**
## ID
 [SFB-81](https://binnenland.atlassian.net/browse/SFB-81)

 ## Description

 This PR  will make it possible to use feature flags for the application. Example in the future we want a `codelijsten` in the application but we do not want to show it yet on our online application.

 ## Type of change

 - [ ] Bug fix
 - [x] New feature
 - [ ] Breaking change
 - [ ] Other

 ## How to test

Go to `config/environment.js` and change the boolean under `featureFlags` to false. now you will see that the builder/validation switch will appear or dissappear.

 ## Links to other PR's

 - /

### CODE REVIEW

**Comment by JonasVanHoof** [comment]
How should you add a feature flag?

1. If the feature you want to hide behind a flag is under a `route` that does not have the service injected
    * inject the service `features` into the route
    * return the service in the model
2. In your `.hbs` file you can use `this.model.features.<yourAmazingfeatureName>` 
3. In your `.js` file you can access it by writing `this.model.features.isEnabled('<yourAmazingfeatureName>')`
4. Make sure the applicaiton is refreshed and you will see the result depending on the config

File: app/controllers/formbuilder/edit.js [hunk 1]
```diff
@@ -89,6 +89,9 @@ export default class FormbuilderEditController extends Controller {
   });
 
   setupPreviewForm = restartableTask(async () => {
+    if (this.model.features.isEnabled('showLogWhenPreviewIsBeingrefreshed')) {
+      console.log(`Refreshing preview`);
+    }
```
Comment (JonasVanHoof) [inline app/controllers/formbuilder/edit.js:94]: Remove*

File: app/templates/formbuilder/edit.hbs [hunk 2]
```diff
@@ -11,14 +11,19 @@
...
+        <AuToggleSwitch
+          @disabled={{or
+            this.setupBuilderForm.isRunning
+            this.setupPreviewForm.isRunning
+          }}
+          @checked={{not this.isShowBuilder}}
+          @onChange={{this.toggleIsAddingValidationToForm}}
+        >Validation</AuToggleSwitch>
+      </div>
+    {{/if}}
```
Comment (JonasVanHoof) [inline app/templates/formbuilder/edit.hbs:26]: Remove the if around it

File: config/environment.js [hunk 3]
```diff
@@ -18,6 +18,10 @@ module.exports = function (environment) {
       // Here you can pass flags/options to your application instance
       // when it is created
     },
+    featureFlags: {
+      showBuilderValidationSwitch: false,
+      showLogWhenPreviewIsBeingrefreshed: false,
```
Comment (JonasVanHoof) [inline config/environment.js:23]: *remove the flags

File: app/routes/formbuilder/edit.js [hunk 4]
```diff
@@ -25,6 +26,7 @@ export default class FormbuilderEditRoute extends Route {
       generatedForm,
       formTtl,
       metaTtl,
+      features: this.features,
```
Comment (wolfderechter) [inline app/routes/formbuilder/edit.js:29]:
```suggestion

```

File: app/templates/formbuilder/edit.hbs [hunk 5]
```diff
@@ -11,14 +11,19 @@
...
   <div class="au-c-main-container__content">
-    <div class="playground-tab">
-      Builder
-      <AuToggleSwitch
-        @disabled={{or this.setupBuilderForm.isRunning this.setupPreviewForm.isRunning}}
-        @checked={{not this.isShowBuilder}}
-        @onChange={{this.toggleIsAddingValidationToForm}}
-      >Validation</AuToggleSwitch>
-    </div>
+    {{#if this.model.features.showBuilderValidationSwitch}}
```
Comment (wolfderechter) [inline app/templates/formbuilder/edit.hbs:14]:
```suggestion
    {{#if this.features.showBuilderValidationSwitch}}
```

**Review by wolfderechter — COMMENTED** [review]
I think things can be a bit simpler by just using `this.features...` instead of the `this.model.features`

**Comment by JonasVanHoof** [comment]
> I think things can be a bit simpler by just using `this.features...` instead of the `this.model.features`

I agree that we should just inject the service in the component or item we are adding the feature to. This limits the scope.

For this PR: will remove the examples and just leave the adding o the package and in the `environment.js` add `featureFlags: {}`

**Comment by JonasVanHoof** [comment]
> > ```
> > 3. In your `.js` file you can access it by writing `this.model.features.isEnabled('<yourAmazingfeatureName>')`
> > ```
> 
> We can also just use `this.features.get('<yourAmazingfeatureName>')` for js

I you read the code `this.features.get('<yourAmazingfeatureName>')` it is less descriptive IMO than `this.features.isEnabled('<yourAmazingfeatureName>')` but that's preference I think. In the future when we start adding implementations behind feature flags we will flashback on this on how to work with it best.

**Pull request merged**